devclean.sh

```diff
@@ -61,6 +59,45 @@ usage_error () {
   exit 1
 }
 
+# remove_directory function
```

Comment:
Sorry for being difficult, but I don't like this solution either. Why do we need to give the users a choice for what directories to delete? If users want to specify some custom directory to delete, they can just delete it! Get rid of these options entirely, simply hard-code the directories and files that need to be deleted. No need to bother with input arguments and checking for wildcards: all of this complexity just leads to *more* potential failure modes.

Reply:
I'd like to get a better understanding what are the features that are desired vs. what was coded earlier. Let me list the old/existing behavior and a new behavior below; please let me know what is to be added to the existing features, and what are excessive. I did refer to the issues mentioned in [#1073](https://github.com/ufs-community/ufs-srweather-app/issues/1073), such as removing directories outside the SRW tree, or those mentioned in the links: [this Quora link](https://www.quora.com/What-are-some-crazy-rm-rf-stories-you-have-heard-abouturl) , https://github.com/ValveSoftware/steam-for-linux/issues/3671 , which included raising issues of having empty variables or wildcards.

 The old behavior: 
- User requests what type of cleaning is to be done using arguments to a script, which could be to do the following
     - remove build directory only;
     - remove all build artifacts (except for built conda directory + conda_loc file); 
     - remove all checked-out external repositories/submodules;
     - remove conda directory and conda_loc;
-  When type of cleaning is determined based on the arguments, the script checks if a directory or directories exists and then removes them. This is done for all but conda + conda_loc (check for directory existence not done).

New behavior, type of cleaning to be done using similar arguments:

- Checks whether a directory exists;
- Check whether a directory is immediately under the expected SRW tree (ask to confirm a remove request if it is not)
- Check whether a directory name contains wildcards, such as " [  ] | ? + *" or a space, do not remove if it does and ask the user to do it from a command line

Reply:
Let me clarify some of the background of my comments: I did not know about the existence of this script for some time, and when I found out about it, I immediately noticed some unsafe behavior as I've mentioned in my open issue. On thinking about it further as I typed up that issue, it struck me that this script is completely unnecessary: all it does is remove directories in known locations. But since you insisted it is still necessary to support users with no advanced knowledge, then I can be happy with a simple script that removes the default directory locations, even if I personally think it's unnecessary.

This functionality that you have added in this PR has greatly increased the complexity of the script, and offers countless more potential failure modes due to introducing complex/advanced regex that purports to remove certain potentially dangerous characters. Furthermore, **the functionality you are attempting to preserve (deleting custom build, install, etc. directories) only applies to advanced users anyway, who should be able to follow very simple instructions to remove the directories if they want a clean build.** These are directories that they will already have had to specify explicitly at build time (assuming that they are different than the default locations), and it is functionality that is not even documented in the users guide!

Simple is better here, especially if you are talking about a tool for supporting novice users. I propose that this script, if it must continue to exist, simply deletes the default locations of these directories (relative to the script itself), and does not attempt to allow the user to input custom locations for these directories.

Reply:
@mkavulich - thank you for your comments. This devclean.sh script was added in the late 2022: https://github.com/ufs-community/ufs-srweather-app/pull/415

This devclean.sh was created to accompany a devbuild.sh script, which accept custom locations for the build artifacts that could be specified, and have variables INSTALL_DIR, CONDA_DIR, and BIN_DIR available. It is also included in the SRW Documentation
https://ufs-srweather-app.readthedocs.io/en/develop/UsersGuide/Reference/FAQ.html#how-can-i-clean-up-the-srw-app-code-if-something-went-wrong-during-the-build

If these variables are not set by the user, they accept default locations, such as under main SRW tree.
So if the original code looks good enough, I could only proposed to add a check for the existence of a CONDA_DIR before removing it, when conda cleaning is requested.

Reply:
> It is also included in the SRW Documentation

I am talking specifically about the customization options for INSTALL_DIR, etc. This is advanced behavior that is only documented in the script itself, so I do not expect anyone to be using it outside of advanced users.

> if the original code looks good enough

Do you mean the code before your changes here? I have already expressed my concerns about that in Issue #1073 that started this discussion. Otherwise I'm not sure what you mean.

I don't think you are addressing my primary concern here: There needs to be a good reason for including all this complexity and potential failure modes in a script, especially one explicitly designed for novice users, and I just do not see that reason at all. Why include all of this for a use case that I can't imagine anyone ever needing? You are implying that there are users out there who are advanced enough to have a very specific use case which requires customizing their build, install, or conda directories, but these same hypothetical users are not advanced enough to follow simple instructions that they should remove those same directories to clean up an existing build? More code is not always better, and I don't understand the argument against replacing this now-200+ line script with a single sentence in the users guide.

Reply:
A simplified version of devclean.sh has been included in a new PR1100, https://github.com/ufs-community/ufs-srweather-app/pull/1100 , and will be taken out of this PR1089

Reply:
Please remove lines 55 - 93.